text/0265-event-vision.md

```diff
@@ -0,0 +1,72 @@
+# Event basics
+
+## Motivation
+
+The introduction of Events has been contentious, so we want to write down and agree on a few basics.
+
+### What are events?
+
+Events are a type of log which has a required event name and a specific structure which is implied by that event name.
```

Comment:
I think this has been the core of the contention. I believe this was a bad decision, to conflate logs and events. Even though events can be viewed as syntactically equivalent to logs, the purpose seems quite different. 

IMO, events should be used for semantically defined _events_, as well as custom user-defined business-events, and not viewed from a user perspective as being just "a type of log". This makes the distinction between logs and events way to fuzzy, and will lead to user confusion.

I know I have argued this in the past and lost the argument, but I do think it's a mistake to treat events as logs, when from a user perspective, they should have a different semantic purpose.

Reply:
I agree, this wording is the problem

> Events are a type of log which has a required event name

The reason that `events` are being treated as a LogRecord was
- to avoid introducing a new transport / protobuf
- Span Events didn't (and now won't) support nested attributes

As John mentions, events are NOT "logs" (they are only transported via Logs)

Reply:
Maybe we need to take a step back and see if folks have the same understanding of "logs".

1. https://opentelemetry.io/docs/concepts/signals/logs/ "A log is a timestamped text record, either structured (recommended) or unstructured, with optional metadata."
2. https://www.crowdstrike.com/cybersecurity-101/observability/three-pillars-of-observability/ "Every activity that happens in your applications and systems generates an [event log](https://www.crowdstrike.com/cybersecurity-101/observability/event-log/) with details that include things like a timestamp, event type, and machine or user ID. Whether written in plain text or binary, [structured or unstructured](https://www.crowdstrike.com/cybersecurity-101/observability/structured-logging/), log data contains text and metadata that provide a level of granularity that’s valuable for debugging and gaining insights into system events and errors."
3. https://www.techtarget.com/searchitoperations/tip/The-3-pillars-of-observability-Logs-metrics-and-traces "Logs are files that record events, warnings and errors as they occur within a software environment. Most logs include contextual information, such as the time an event occurred and which user or endpoint was associated with it."

I can start with my understanding:

* I think the opentelemetry.io description ("is a ... text record") is not what OpenTelemetry envisions, it doesn't have to be limited to "text" at all.
* I think the TechTarget description is not what OpenTelemetry envisions, logs are not limited to files.
* I think the CrowdStrike description is better aligned with what we want to have in OpenTelemetry.

Reply:
my POV:
- logs (unstructured or semi structured blobs of text with no semantics) is a legacy, almost everything emits them, but we don't people to use them
- events (structured things with documented semantics) is the future, we want everyone to start using them

Existing logging facades usually do something in between - you can use logging APIs in a structured or unstructured way.

By saying that events are logs v2, we're trying to move everyone to a structured data. We're saying that OTel will (where necessary) provide the better-than-logs thing. We'll discourage using logs for new projects based on otel. But we'll stay compatible with unstructured logs.

If we try to separate logs and events:
- we'll keep unstructured logs as one of the things we support 
- we'll keep existing level of confusion among ourselves and users (do I use log or event for that thing)?
- we'll not have a reasonable story for structured-semantical-documented-things you create with traditional logging

It seems all the difference about logs and events is individual perception. Since we have different perceptions within this community, we'd fail to explain the difference to users.

Reply:
If events are "logs v2", I think we should not transport them over the same channel as "logs v1". They should be separated, so as not to cause confusion between the two. How is one to know if a given log record is a "v1 log" or a v2 log (i.e. "event")? You can easily use "v1 logs" structure to create precisely identical payloads to a "v2 log", and hence there is no practical way to distinguish them.

I would absolutely support events being "logs v2", and I actually think this is an excellent idea, but as long as they are co-mingled in the transport with "logs v1" records, I think we will never be able to resolve how they differ.

Reply:
logs v2 support all that logs v1 support from data structure standpoint + more - is there any controversy in it?

We have a discriminating property - presence of the `event.name` attribute, we can in theory create other discriminating properties if that's not enough for some reason. 

if you can use traditional logs to create properly structured and named event, then such logs are already logs v2  - what's wrong with this?

Reply:
to be more precise: Events API is Logs API v2, and we happen to have data model that's already ~perfect~ compatible with both.

Reply:
If you can already use a traditional log API to generate logs v2, then what value are we adding? Events are nothing but semantic conventions for logs, then, and I think we should drop the name "events" entirely. Unless our events are really something unique from existing structured logs, let's skip them altogether.

Reply:
> Unless our events are really something unique from existing structured logs, let's skip them altogether.

+1

Reply:
events are both structured and named (and the expectation is that the name implies the structure)

Reply:
> If you can already use a traditional log API to generate logs v2, then what value are we adding? 

traditional log API prioritizes string bodies and varies in ability to record context.

You'd have a subpar experience reporting structured bodies using SLF4J, Log4J, .NET ILogger, and many other loggers that support structured logging. You'll have no luck with loggers that don't support it such as JUL and many others.

With structured loggers, we'd need some subtle conventions for logging APIs to achieve it. E.g. potential approach with slf4j

```
logger.atInfo()
 .addKeyValue("event.name", "com.foo.my-event-id")
 .addKeyValue("otel.log.body", myEventBody) // `otel.log.body` property is used as log record body by bridge api
 .log("something important") 
```

Events API/Log v2 API is necessary to provide the API capable of defining all sorts of events consistently in OTel. If we don't have it, we'll need to solve structured logging for each external logging facade we need to support.

Some logging facade (like .NET ILogger)  may change to accommodate events better, but most of them probably won't. 

---------------------------

TL;DR: call it Event API or Logging API (v2), we need to have some API in OTel that allows any otel-api user to create structured named events (aka logs).

A different name for this API is helpful so we don't need to explain that it's not string-based log with questionable structure.

Reply:
How is this conversation resolved? It seems like dissenting voices are just getting steamrolled here.

Reply:
I probably resolved it based on the comment above because our events really are something unique from existing structured logs: https://github.com/open-telemetry/oteps/pull/265#discussion_r1736401546

Unfortunately the discussions around events have tended to spin out into many different directions which makes it hard to know when they are resolved, or if they are only partially resolved then what portions still need to be resolved, which is why I've requested https://github.com/open-telemetry/oteps/pull/265#issuecomment-2317929226

> since some of the discussions are quite long, instead of unresolving it may help instead to start a new discussion with a summary of remaining discussion points (along with a pointer to the prior discussion for full context)

Reply:
Gotcha, thanks for the link. I had missed that comment.